### PR TITLE
Bin pack distcp-ng work units.

### PR DESCRIPTION
**Pull request by ibuenros**
Implemented bin packing for distcp-ng. Made it generic enough to be usable by other Gobblin flows.

### CODE REVIEW

**Comment by ibuenros** [comment]
@pcadabam can you review?

**Comment by pcadabam** [comment]
Few comments
- I think bin packing should be in gobblin-api/gobblin.source.workunit

File: gobblin-utility/src/main/java/gobblin/util/binpacking/WorstFitDecreasingBinPacking.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+ *   * Compute the minimum number of {@link MultiWorkUnit}s needed, and create them.
+ *   * For each work unit, find the {@link MultiWorkUnit} with the largest space available, if it fits, add it there,
+ *     otherwise, create a new {@link MultiWorkUnit} and add the work unit there.
+ * </p>
+ */
+public class WorstFitDecreasingBinPacking {
+
+  public static final String TOTAL_MULTI_WORK_UNIT_WEIGHT = "binpacking.multiWorkUnit.totalWeight";
+
+  public List<WorkUnit> pack(List<WorkUnit> workUnitsIn, Weighter weighter, long maxWeigthPerUnit) {
```
Comment (pcadabam) [inline gobblin-utility/src/main/java/gobblin/util/binpacking/WorstFitDecreasingBinPacking.java]: `WorstFitDecreasingBinPacking` should implement `WorkUnitBinPacking` interface. The interface can have a `public List<WorkUnit> pack(List<WorkUnit> workUnitsIn, Weighter weighter, long maxWeigthPerUnit)` method

File: gobblin-utility/src/test/java/gobblin/util/binpacking/WorstFitDecreasingBinPackingTest.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import gobblin.source.workunit.MultiWorkUnit;
+import gobblin.source.workunit.WorkUnit;
+
+
+public class WorstFitDecreasingBinPackingTest {
+
+  public static final String WEIGHT = "weight";
+
+  @Test
+  public void testPack()
```
Comment (pcadabam) [inline gobblin-utility/src/test/java/gobblin/util/binpacking/WorstFitDecreasingBinPackingTest.java]: As a general note it is better to break down a test class into logical test methods. The idea is to have as few Asserts in a test method as possible. This is makes it easy to debug when a test fails.

**Comment by zliu41** [comment]
This may overlap with `KafkaWorkUnitPacker`, is it possible to consolidate these two?

**Comment by ibuenros** [comment]
@zliu41 yes, in fact I pretty much copied it from `KafkaWorkUnitPacker`. However, the algorithm is slightly different (the Kafka bin packer has bounded number of workunits, this algorithms bounds the total weight of each work unit), so I didn't want to change what the `KafkaWorkUnitPacker` uses.
I can either change kafka bin packing to use this new algorithm, or implement the other algorithm as well, although for the latter, I would prefer to do that in a later PR.

**Comment by ibuenros** [comment]
@pcadabam addressed comments. Can you take another quick look? Thanks!

File: gobblin-api/src/main/java/gobblin/source/workunit/WorkUnitBinPacker.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * A bin packing algorithm for packing {@link WorkUnit}s into {@link MultiWorkUnit}s.
+ */
+public interface WorkUnitBinPacker {
+
+  /**
+   * Packs the input {@link WorkUnit}s into {@link MultiWorkUnit}s.
+   */
+  public List<WorkUnit> pack(List<WorkUnit> workUnitsIn, WorkUnitWeighter weighter);
```
Comment (pcadabam) [inline gobblin-api/src/main/java/gobblin/source/workunit/WorkUnitBinPacker.java]: Add `@param` in java doc

**Comment by pcadabam** [comment]
LGTM with that one comment about javadoc

**Pull request merged**